substrate/frame/sassafras/src/lib.rs

```diff
@@ -0,0 +1,1079 @@
+
+		let accumulator = hashing::blake2_256(&buf);
+		RandomnessAccumulator::<T>::put(accumulator);
+	}
+
+	// Deposit next epoch descriptor in the block header digest.
+	fn deposit_next_epoch_descriptor_digest(desc: NextEpochDescriptor) {
+		let item = ConsensusLog::NextEpochData(desc);
+		let log = DigestItem::Consensus(SASSAFRAS_ENGINE_ID, item.encode());
+		<frame_system::Pallet<T>>::deposit_log(log)
+	}
+
+	// Initialize authorities on genesis phase.
+	//
+	// Genesis authorities may have been initialized via other means (e.g. via session pallet).
+	//
+	// If this function has already been called with some authorities, then the new list
+	// should match the previously set one.
+	fn genesis_authorities_initialize(authorities: &[AuthorityId]) {
+		let prev_authorities = Authorities::<T>::get();
+
+		if !prev_authorities.is_empty() {
+			// This function has already been called.
+			if prev_authorities.as_slice() == authorities {
+				return
+			} else {
+				panic!("Authorities were already initialized");
+			}
+		}
+
+		let authorities = WeakBoundedVec::try_from(authorities.to_vec())
+			.expect("Initial number of authorities should be lower than T::MaxAuthorities");
+		Authorities::<T>::put(&authorities);
+		NextAuthorities::<T>::put(&authorities);
+	}
+
+	// Method to be called on first block `on_initialize` to properly populate some key parameters.
+	fn post_genesis_initialize(slot: Slot) {
+		// Keep track of the actual first slot used (may not be zero based).
+		GenesisSlot::<T>::put(slot);
+
+		// Properly initialize randomness using genesis hash and current slot.
+		// This is important to guarantee that a different set of tickets are produced for:
+		// - different chains which share the same ring parameters and
+		// - same chain started with a different slot base.
+		let genesis_hash = frame_system::Pallet::<T>::parent_hash();
+		let mut buf = genesis_hash.as_ref().to_vec();
+		buf.extend_from_slice(&slot.to_le_bytes());
+		let randomness = hashing::blake2_256(buf.as_slice());
+		RandomnessAccumulator::<T>::put(randomness);
+		let next_randoness = Self::update_epoch_randomness(1);
+
+		// Deposit a log as this is the first block in first epoch.
+		let next_epoch = NextEpochDescriptor {
+			randomness: next_randoness,
+			authorities: Self::next_authorities().into_inner(),
+			config: None,
+		};
+		Self::deposit_next_epoch_descriptor_digest(next_epoch);
+	}
+
+	/// Current epoch information.
+	pub fn current_epoch() -> Epoch {
+		let index = EpochIndex::<T>::get();
+		Epoch {
+			index,
+			start: Self::epoch_start(index),
+			length: T::EpochLength::get(),
+			authorities: Self::authorities().into_inner(),
+			randomness: Self::randomness(),
+			config: Self::config(),
+		}
+	}
+
+	/// Next epoch information.
+	pub fn next_epoch() -> Epoch {
+		let index = EpochIndex::<T>::get()
+			.checked_add(1)
+			.expect("epoch indices will never reach 2^64 before the death of the universe; qed");
```

Comment:
```suggestion
		let index = EpochIndex::<T>::get() + 1;
```

Easier and less code ;)